src/PostgREST/ContentTypes.hs

```diff
@@ -0,0 +1,66 @@
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE DeriveGeneric         #-}
+{-# LANGUAGE DuplicateRecordFields #-}
+
+module PostgREST.ContentTypes
+  ( ContentType(..)
+  , toHeader
+  , toMime
+  , decodeContentType
+  ) where
+
+import qualified Data.ByteString          as BS
+import qualified Data.ByteString.Internal as BS (c2w)
+
+import Network.HTTP.Types.Header (Header, hContentType)
+
+import Protolude
+
+-- | Enumeration of currently supported response content types
+data ContentType
+  = CTApplicationJSON
+  | CTSingularJSON
+  | CTTextCSV
+  | CTTextPlain
+  | CTOpenAPI
+  | CTUrlEncoded
+  | CTOctetStream
+  | CTAny
+  | CTOther ByteString
+  deriving (Eq)
+
+-- | Convert from ContentType to a full HTTP Header
+toHeader :: ContentType -> Header
+toHeader ct = (hContentType, toMime ct <> charset)
+  where
+    charset = case ct of
+      CTOctetStream -> mempty
+      CTOther _     -> mempty
+      _             -> "; charset=utf-8"
+
+-- | Convert from ContentType to a ByteString representing the mime type
+toMime :: ContentType -> ByteString
+toMime CTApplicationJSON = "application/json"
+toMime CTTextCSV         = "text/csv"
+toMime CTTextPlain       = "text/plain"
+toMime CTOpenAPI         = "application/openapi+json"
+toMime CTSingularJSON    = "application/vnd.pgrst.object+json"
+toMime CTUrlEncoded      = "application/x-www-form-urlencoded"
+toMime CTOctetStream     = "application/octet-stream"
+toMime CTAny             = "*/*"
+toMime (CTOther ct)      = ct
+
+-- | Convert from ByteString to ContentType. Warning: discards MIME parameters
+decodeContentType :: BS.ByteString -> ContentType
+decodeContentType ct =
+  case BS.takeWhile (/= BS.c2w ';') ct of
+    "application/json"                  -> CTApplicationJSON
+    "text/csv"                          -> CTTextCSV
+    "text/plain"                        -> CTTextPlain
+    "application/openapi+json"          -> CTOpenAPI
+    "application/vnd.pgrst.object+json" -> CTSingularJSON
+    "application/vnd.pgrst.object"      -> CTSingularJSON
+    "application/x-www-form-urlencoded" -> CTUrlEncoded
+    "application/octet-stream"          -> CTOctetStream
+    "*/*"                               -> CTAny
+    ct'                                 -> CTOther ct'
```

Comment:
Possible refactor: Change `toMime` and `decodeContetType` to instances of `Show` and `Read`.

Would that be more ~~pythonic~~ haskell-ish?

Reply:
By my understanding our uses of `Show` and `Read` in other places are a bit of an abuse... See e.g. "How do I write manual Show instances if show isn't provided?" in https://hackage.haskell.org/package/protolude . 

The [Haskell docs say](https://hackage.haskell.org/package/base-4.15.0.0/docs/Text-Show.html): "The result of show is a syntactically correct Haskell expression containing only constants," That's not the case with the contenttypes here, for example.

So I think using simple functions that return Text or ByteStrings is fine! We should probably also refactor the other `Show` instances and avoid importing `GHC.Show`.

Reply:
I read quite a bit about it and it's a complex topic. I had a long post dissecting it bit-by-bit (assuming I'd argue to still use `Show` / `Read`) and then turned out come... to the same conclusion. Not exactly for the same reasons, but avoiding `GHC.Show` seems like a good idea.

In that case, I suggest to rename the functions slightly - if we decide to use the pattern above. Will comment there.